### PR TITLE
Add helm chart to install trivy in server mode.

### LINKED ISSUE #688: [FR] provide helm chart for trivy server
Hi,

I am using Harbor but didn't integrate with trivy.
Because I experienced harbor performance issue. (actually it was clair anyway)

I just want to deploy trivy server 
for any trivy commands can use it.

can you provide helm chart or operator?

key features are
1) save db at PV
2) schedule db update
3) provide endpoint

### PR DESCRIPTION
**Pull request by czunker**
Fixes #688

Signed-off-by: Christian Zunker <christian.zunker@codecentric.cloud>

### CODE REVIEW

**Comment by knqyf263** [comment]
@czunker Thanks for your great contribution!
@danielpacak Could you help me with the review? It should be easy to integrate into Starboard.

File: helm/trivy/templates/statefulset.yaml
Comment (danielpacak) [inline helm/trivy/templates/statefulset.yaml:52]: We need to double check, but I played with Trivy server run in the cluster and I had to specify `0.0.0.0:8080` to accept connection from client pods. See https://github.com/aquasecurity/starboard/blob/master/deploy/trivy-server/03-trivy-server.deployment.yaml#L32

File: helm/trivy/templates/statefulset.yaml
Comment (danielpacak) [inline helm/trivy/templates/statefulset.yaml:57]: That's the name of env recognized by Trivy:

```suggestion
            - name: "TRIVY_DEBUG"
```

File: helm/trivy/values.yaml
Comment (danielpacak) [inline helm/trivy/values.yaml:42]: I'd not expose it as Helm value. See my previous comment where this value is used.

File: helm/trivy/values.yaml
Comment (danielpacak) [inline helm/trivy/values.yaml:39]: Does not take effect

File: helm/trivy/values.yaml
Comment (danielpacak) [inline helm/trivy/values.yaml:82]: I'd use the default port of trivy server 4954

File: helm/trivy/Chart.yaml
Comment (danielpacak) [inline helm/trivy/Chart.yaml:8]:
```suggestion
  - trivy
```

File: helm/trivy/README.md
Comment (danielpacak) [inline helm/trivy/README.md:29]: scanner adapter -> Trivy server

File: helm/trivy/README.md
Comment (danielpacak) [inline helm/trivy/README.md:55]: Takes no effect

File: helm/trivy/README.md
Comment (danielpacak) [inline helm/trivy/README.md:57]: Not used by Trivy server

File: helm/trivy/README.md
Comment (danielpacak) [inline helm/trivy/README.md:66]: I'd use 4954 instead of 8080 to be consistent with trivy CLI

File: helm/trivy/templates/NOTES.txt
Comment (danielpacak) [inline helm/trivy/templates/NOTES.txt:1]:
```suggestion
You should be able to access Trivy server installation within
```

File: helm/trivy/templates/statefulset.yaml
Comment (danielpacak) [inline helm/trivy/templates/statefulset.yaml:2]: Did you consider using Deployment with emptyDir volumes? I know that you're inspired by aquasecurity/harbor-scanner-trivy, but Harbor adapter is using Trivy in standalone mode. That was the reason to use StatefulSets, but in this case Deployment should work.

File: helm/trivy/values.yaml
Comment (danielpacak) [inline helm/trivy/values.yaml:31]: For aquasecrity/harbor-scanner-trivy image we created user scanner(id=1000). It's not the case for Trivy image. So maybe you can use nobody(id=65534)

File: helm/trivy/values.yaml
Comment (danielpacak) [inline helm/trivy/values.yaml:75]: I think we don't use this anywhere. The server is not pulling image layers.

File: helm/trivy/README.md
Comment (danielpacak) [inline helm/trivy/README.md:64]: Not used.

File: helm/trivy/templates/statefulset.yaml
Comment (danielpacak) [reply helm/trivy/templates/statefulset.yaml:52]: Actually, I've just tested it and it works fine!

**Review by danielpacak — CHANGES_REQUESTED** [review]
👋 @czunker Thanks for this PR. It looks and works great (tested in my cluster). I left a few comments to cleanup a bit descriptors and configs that were only pertinent to Harbor scanner adapter.

File: helm/trivy/templates/statefulset.yaml
Comment (knqyf263) [reply helm/trivy/templates/statefulset.yaml:49]: Yes, Trivy recognizes only `TRIVY_DEBUG`.

File: helm/trivy/values.yaml
Comment (knqyf263) [reply helm/trivy/values.yaml:75]: As @danielpacak said, the server doesn't communicate with a registry.

File: helm/trivy/templates/statefulset.yaml
Comment (czunker) [reply helm/trivy/templates/statefulset.yaml:2]: I understood the comments in a way, that Trivy would pull the database from github.
With the PVC and statefulset I wanted to prevent a pull for the database on every POD restart.

Comment (danielpacak) [reply helm/trivy/templates/statefulset.yaml:2]: @czunker Okey so in case you want to reuse downloaded DB files by configuring the storage class that has the reclaim policy set to `Retain`, having StatefulSet makes sense.  (AFAIK This is not possible with Deployments and the only way is indeed `PersistenceVolumeClaimTemplate`.)

On the other hand we can change the default `podManagementPolicy` of the StatefulSet from `OrderedReady` to `Paraller`, so pods are spun up independently rather than one after the other. There no reason to wait in case of Trivy server. Pods can run and scale up or down independently.

File: helm/trivy/templates/statefulset.yaml
Comment (danielpacak) [inline helm/trivy/templates/statefulset.yaml:60]: Add `trivy.ignoreUnfixed` to values.yaml with short description

File: helm/trivy/templates/statefulset.yaml
Comment (danielpacak) [inline helm/trivy/templates/statefulset.yaml:62]: Add `trivy.skipUpdate` to values.yaml with short description

**Review by danielpacak — CHANGES_REQUESTED** [review]
@czunker Good progress. I pointed out some other things. Please make sure that you addressed the comments from my first review. There's still unresolved discussions. Thank you!

File: helm/trivy/templates/statefulset.yaml
Comment (czunker) [reply helm/trivy/templates/statefulset.yaml:2]: I changed the `podManagementPolicy` and added a short note in the Readme.

**Comment by czunker** [comment]
Hi @danielpacak ,

hope I didn't miss a comment this time.

File: helm/trivy/templates/statefulset.yaml
Comment (danielpacak) [inline helm/trivy/templates/statefulset.yaml:49]: Even though we do not expose cache directory as Helm value, we have to set it as ENV here. When I tested it pods were failing because we're running Trivy server as non root user (`nobody`), but it's trying to read and write to the `/.cache` directory.

```yaml
env:
  - name: "TRIVY_CACHE_DIR"
    value: "/home/scanner/.cache/trivy"
```

File: helm/trivy/values.yaml
Comment (danielpacak) [inline helm/trivy/values.yaml:53]: I think we're missing description of the `gitHubToken` property, which is referred to by the Secret.

File: helm/trivy/templates/service.yaml
Comment (danielpacak) [inline helm/trivy/templates/service.yaml:7]: By default this service has `spec.sessionAffinity: None`, which might cause inconsistency errors when we scale up StatefulSet replicas. I was actually able to see such errors in my env after scaling to 3 replicas.

Basically we have to make sure that all requests related to scanning a particular image are routed to the same pod, which has its own file system cache (PVC -> PV).

The following sketch diagram depicts the problem that I'm trying to describe.

trivy-server-session-affinity

One way to fix it is to change `spec.sessionAffinity: None` to `spec.sessionAffinity: ClientIP`, but we have to double check how does that impact scalability. /cc @knqyf263

**Review by danielpacak — CHANGES_REQUESTED** [review]
Great progress @czunker Just a few more things to change and consider before we can merge the PR please.

File: helm/trivy/values.yaml
Comment (czunker) [reply helm/trivy/values.yaml:53]: I think I missed to delete. In another comment mentioned the github token isn't needed to pull the database.

File: helm/trivy/templates/statefulset.yaml
Comment (czunker) [reply helm/trivy/templates/statefulset.yaml:49]: Added this.

File: helm/trivy/templates/service.yaml
Comment (czunker) [reply helm/trivy/templates/service.yaml:7]: Added

File: helm/trivy/values.yaml
Comment (danielpacak) [reply helm/trivy/values.yaml:53]: Sorry for that @czunker You're right. I saw the Secret referring to the GH token and I mixed up things. Indeed we don't need the Secret and the GH token for Trivy client-server

**Review by danielpacak — APPROVED** [review]

File: helm/trivy/templates/service.yaml
Comment (knqyf263) [reply helm/trivy/templates/service.yaml:7]: I've supported Redis and it will address this issue.
https://github.com/aquasecurity/fanal/pull/143

Should we wait for the Redis support? Or, can we merge this PR once and replace PV with Redis in separate PR? What do you think, @danielpacak?

**Review by knqyf263 — APPROVED** [review]

File: helm/trivy/templates/service.yaml
Comment (knqyf263) [reply helm/trivy/templates/service.yaml:7]: It will be done in the subsequent PR.

**Comment by knqyf263** [comment]
Thanks for the great contribution! @czunker

File: helm/trivy/values.yaml
Comment (knqyf263) [reply helm/trivy/values.yaml:53]: I overlooked this conversation, but I was talking about `insecure`, not `githubToken`. I meant `insecure` was unnecessary, but  `githubToken` is necessary for GitHub rate limiting.

Comment (czunker) [reply helm/trivy/values.yaml:53]: Fixed this in PR #769

Comment (danielpacak) [reply helm/trivy/values.yaml:53]: Absolutely. I mixed up pulling layers from registry server by client (does not require GH token) and downloading Trivy DB from GH which is subject to rate limit and we should be able to configure GH token.

**Pull request merged**
